core/apple/src/SourcesApple.kt

```diff
@@ -0,0 +1,121 @@
+
+import kotlinx.cinterop.*
+import platform.Foundation.*
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.darwin.NSUIntegerVar
+import platform.posix.memcpy
+import platform.posix.uint8_tVar
+import kotlin.concurrent.Volatile
+
+/**
+ * Returns an input stream that reads from this source. Closing the stream will also close this source.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.asStream
+ */
+public fun Source.asNSInputStream(): NSInputStream = SourceNSInputStream(this)
+
+@OptIn(InternalIoApi::class, UnsafeNumber::class)
+private class SourceNSInputStream(
+    private val source: Source,
+) : NSInputStream(NSData()) {
+
+    private val isClosed: () -> Boolean = when (source) {
+        is RealSource -> source::closed
+        is Buffer -> {
+            { false }
+        }
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    @Volatile
+    private var status = NSStreamStatusNotOpen
+    private var error: NSError? = null
+        set(value) {
+            status = NSStreamStatusError
+            field = value
+        }
+
+    private var pinnedBuffer: Pinned<ByteArray>? = null
+
+    override fun streamStatus() = if (isClosed()) NSStreamStatusClosed else status
+
+    override fun streamError() = error
+
+    override fun open() {
+        if (status == NSStreamStatusNotOpen) {
+            status = NSStreamStatusOpen
+        }
+    }
+
+    override fun close() {
+        status = NSStreamStatusClosed
+        pinnedBuffer?.unpin()
+        pinnedBuffer = null
+        source.close()
+    }
+
+    override fun read(buffer: CPointer<uint8_tVar>?, maxLength: NSUInteger): NSInteger {
+        try {
+            if (isClosed()) throw IOException("Underlying source is closed.")
+            if (status != NSStreamStatusOpen) return -1
+            status = NSStreamStatusReading
+            if (source.exhausted()) {
+                status = NSStreamStatusAtEnd
+                return 0
+            }
+            val toRead = minOf(maxLength.toInt(), source.buffer.size).toInt()
+            val read = source.buffer.readNative(buffer, toRead).convert<NSInteger>()
+            status = NSStreamStatusOpen
+            return read
+        } catch (e: Exception) {
+            error = e.toNSError()
+            return -1
+        }
+    }
+
+    override fun getBuffer(buffer: CPointer<CPointerVar<uint8_tVar>>?, length: CPointer<NSUIntegerVar>?): Boolean {
+        if (source.buffer.size > 0) {
+            source.buffer.head?.let { s ->
+                pinnedBuffer?.unpin()
+                s.data.pin().let {
+                    pinnedBuffer = it
+                    buffer?.pointed?.value = it.addressOf(s.pos).reinterpret()
+                    length?.pointed?.value = (s.limit - s.pos).convert()
+                    return true
+                }
+            }
+        }
+        return false
+    }
+
+    private fun Buffer.readNative(sink: CPointer<uint8_tVar>?, maxLength: Int): Int {
```

Comment:
Should we iterate over segments here while we not yet read `maxLength`?

Reply:
For `Source` reads, the [JVM implementation](https://github.com/Kotlin/kotlinx-io/blob/03845eaa0f62c4d54fc5f3ea0559e0176c2c5fa5/core/common/src/Buffer.kt#L326-L344) doesn't iterate over segments. [This test](https://github.com/Kotlin/kotlinx-io/pull/174/files#diff-8af6b3dce7132e2e2dba9eb3e0a373944f872248e24c670243a0e6bdbf5a665dR40) specifically tests for this behavior, which I initially borrowed from [this Okio test](https://github.com/square/okio/blob/bec51e206c7b95a01a9ccbe9eff126d28f2fac2b/okio/src/jvmTest/kotlin/okio/OkioTest.kt#L108). I don't see this same test in kotlinx-io though.